pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java

```diff
@@ -366,4 +366,74 @@ public boolean equals(Object obj) {
       return _column.equals(that._column) && _dataType == that._dataType;
     }
   }
+
+  /**
+   * Row Index based APIs for Single Value columns
+   */
+
+  public int getSVIntAtIndex(final String column, final int index) {
```

Comment:
It is unclear to me as to why we need to fetch values row-by-row for distinct, I'd rather avoid exposing this api to avoid function call per row? Also, there seems to be code duplication, `getSVIntAtIndex` should have just called `getIntValuesForSVColumn`?

Reply:
At the end to judge uniqueness, we need to look at a row. So simply looking at the return value (which is an array containing all the projected values for the column) from getIntValuesForSVColumn will not help. We should look at each cell from each such array together with the values for other projection column arrays to check if row has already been stored in hashset and store it if now

That's why I introduced the index based APIs.

I see your point though w.r.t making a function call per row... what we can do is use the existing APIs and fetch the values(the array) and then iterate over them to build row, check for existence and store if new. We can avoid the function call overhead this way...